### ISSUE TITLE
aria-activedescendant for empty list

### ISSUE THREAD

**Issue by Jakooth** [validator/validator#557]
Empty list with `aria-activedescendant` are reported as:
"Bad value  for attribute aria-activedescendant on element div: An ID must not be the empty string."
Example:

    <div role="listbox" aria-activedescendant></div>

This is in contradiction with this example:
https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox.html

>  "If the listbox element does not contain any options, `aria-activedescendant` does not have a value."

Customers are testing our applications for accessibility compliance and this is often popping up and we have to explain this is not an issue each time. Would be nice to fix.

Thanks!

**Comment by cvrebert** [comment]
The actual spec says the value must be an ID reference, and makes no mention of also allowing empty values:
https://www.w3.org/WAI/PF/aria/states_and_properties#aria-activedescendant

**Comment by nschonni** [comment]
Related on the practices repo https://github.com/w3c/aria-practices/issues/429

**Comment by sideshowbarker** [comment]
> Related on the practices repo w3c/aria-practices#429

Thanks—from there I see a link to https://github.com/w3c/aria/issues/501 (_“Allow null value of aria-activedescendent when element does not have DOM focus“_) so I guess this is yet another issue where it’s known that the WAI-ARIA Authoring Practices document is using markup that doesn’t conform to the actual ARIA spec requirements.

It’s really disappointing that the group—by not bothering to clearly indicate in the WAI-ARIA Authoring Practices document all the places where it has examples suggesting markup that doesn’t conform to the ARIA spec—is continuing to knowingly confuse web developers/authors and waste their time (and cause issues like this one to get raised).

**Comment by stevefaulkner** [comment]
in fairness it does say on the example cited:

> This example page is incomplete; it is not ready for review. When ready,
> feedback will be welcome in issue 123.
> <https://github.com/w3c/aria-practices/issues/123>

https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox.html

--

Regards

SteveF
Current Standards Work @W3C
<http://www.paciellogroup.com/blog/2015/03/current-standards-work-at-w3c/>

**Comment by sideshowbarker** [comment]
> in fairness it does say on the example cited:
>
> > This example page is incomplete; it is not ready for review. When ready,
> > feedback will be welcome in issue 123.
> > <https://github.com/w3c/aria-practices/issues/123>
>
> https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox.html

That doesn’t make anywhere near clear enough for readers. That statement should set apart with some formatting that gives it very strong emphasis—e.g., with some kind warning indicator/logo, and in bold, etc.

It should also very explicitly say something like, “The following example is invalid in that it doesn’t conform to the current requirements in the ARIA specification.” Just saying it’s “incomplete” doesn’t make that clear.

And it shouldn’t be necessary for me or anybody else to have to explain to a group working on authoring guides why their guides should be very clear about things like the above (to prevent their guides from confusing thousands of developers/authors and wasting their time).

And especially if a group makes a non-normative guide for using a particular technology and **is the same group making the normative spec** for the technology, nobody should need to explain to them how important it is they make their guidance actually conform to the requirements in their own spec.

**Comment by sideshowbarker** [comment]
Closing here because there’s no further action I can take on this. If/when the spec gets updated, a new issue should be raised. In the mean time feel free to comment here further if there’s any new information to add

**Comment by mcking65** [comment]
I believe this may actually be a valid validator bug. My rationale as stated in w3c/aria#501:

[ARIA Section  7.6 State and Property Attribute Processing]() 
states:

> Sometimes states and properties are present in the DOM but have a zero-length string ("") as their value. This is equivalent to their absence. User agents should treat state and property attributes with a value of "" the same as they treat an absent attribute. For supported states and properties, this corresponds to the default value, but if it is a required attribute, it signals an author error, and the implicit value for the role is used.

Since aria-activedescendant is never a required property, shouldn't the empty string be treated as if the attribute is not present?

I don't see anything in the spec that disallows an undefined value for aria-activedescendant or any other non-required property that has an IDREF as its value.

**Comment by sideshowbarker** [comment]
> I believe this may actually be a valid validator bug. My rationale as stated in w3c/aria#501:
>
> [ARIA Section  7.6 State and Property Attribute Processing]() states:

OK, so is spite the title of that section — which would lead most readers to assume it’s defining only UA conformance requirements — I see that it does state some normative requirements for authors, using RFC 2119 _MAY_ and _MUST_ and _SHOULD_. And it also uses those to state the UA requirements; for example, from the text cited in https://github.com/validator/validator/issues/557#issuecomment-335679999:

> User agents _SHOULD_ treat state and property attributes with a value of "" the same as they treat an absent attribute.

But that ARIA spec text cited in https://github.com/validator/validator/issues/557#issuecomment-335679999 doesn’t use any RFC 2119 language to clearly state an authoring requirement allowing authors to use states and properties with a “a zero-length string” as the value.

I mean, given that the spec uses _MAY_ and _MUST_ and _SHOULD_ for other authoring requirements there, I would expect if it intended authors to be allowed to use “a zero-length string” as a value for states and properties it would say that very clearly — like this:

> Authors _MAY_ use a zero-length string as the value for any state or property that isn’t required.

If that’s actually what the authors of the spec intended for it to say, then the fact that it doesn’t say that is a bug in the spec that needs to be fixed.

In the mean time, as far as what the spec does actually provide for authors in a way they can understand clearly — as noted in https://github.com/validator/validator/issues/557#issuecomment-325692179, for `aria-activedescendant`, the information stated at point of use in https://www.w3.org/TR/wai-aria-1.1/#aria-activedescendant is that `aria-activedescendant` requires an **ID reference**  as its value, and at https://www.w3.org/TR/wai-aria-1.1/#valuetype_idref the specs an **ID reference** is this:

> Reference to the ID of another element in the same document

Nowhere there does the spec clearly say to authors that they’re allowed to use “a zero-length string” as the value instead of using a “Reference to the ID of another element in the same document”. So from the point of view of authors reading the spec, it seems quite clear that the `aria-activedescendant` value must be a “Reference to the ID of another element in the same document”.

Anyway, as I’ve stated in other issue discussions, I can’t base the validator behavior on trying to guess what the authors of the ARIA spec might have intended for the spec to say rather than what it actually says. Nor can I base the validator behavior on decisions that the ARIA working group has made but not taken the time to actually normatively define in any specification yet.

**Comment by mcking65** [comment]
Good points. The fact that section 7.6 is directed at UAs is relevant. So, it is clear that empty strings are anticipated and the spec says what UAs should do with them.

Looking for the text that suggests that authors may provide empty strings when a state or property is not required, I find the following.

In [5.2.2 Required States and Properties](http://w3c.github.io/aria/aria/aria.html#requiredState) 
it says:

> Content authors must provide a non-empty value for required states and properties. Content authors must not use the value undefined for required states and properties, unless undefined is an explicitly-supported value of that state or property.

In contrast, section [5.2.3 Supported States and Properties](http://w3c.github.io/aria/aria/aria.html#supportedState) 
says:

> Content authors may provide values for supported states and properties, but need not in some cases where default values are sufficient.

Note the "MAY" and the exclusion where default values are sufficient.

I believe the combination of section 5.2.2 and 5.2.3 with 7.6 languatge does clearly permit authors to provide an empty string for a non-required property and states that doing so is the same as not specifying the attribute. 

It is also worth noting that empty values for properties like aria-labelledby are anticipated in [Section 4.3 of the name calculation spec](http://w3c.github.io/aria/accname-aam/accname-aam.html#mapping_additional_nd_te):

> if computing a name, and the current node has an aria-labelledby attribute that contains at least one valid IDREF, and...

Could the option of providing an empty string for a non-required property be made more apparent to authors within the specification in the way you suggest? Absolutely! However, the lack of such clear direction to authors is not a reason for calling an empty string an author error. BTW, the purpose of the authoring practices is to help authors understand things that are not readily apparent in the specification.

**Comment by sideshowbarker** [comment]
> I believe the combination of section 5.2.2 and 5.2.3 with 7.6 language does clearly permit authors to provide an empty string

I believe it doesn’t — and as I noted in https://github.com/validator/validator/issues/557#issuecomment-335699935, there are plenty of other places in the ARIA spec where the authors use RFC 2119 to clearly state normative document-conformance requirements for authors. But this is not one of those cases. So if in fact the editors of the ARIA spec intended what you’re asserting, then the lack of a simple statement of the requirement is a bug in the spec.

As I wrote in https://github.com/validator/validator/issues/557#issuecomment-335699935 all that’s needed would be a single simple statement of the requirement, like this:

>  Authors MAY use a zero-length string as the value for any state or property that isn’t required.

…or however the ARIA spec editors want word whatever it is they intend the requirement to be.

Without a clearly stated normative requirement like that in the spec, I can’t change the validator behavior for this.

> Could the option of providing an empty string for a non-required property be made more apparent to authors within the specification in the way you suggest? Absolutely! However, the lack of such clear direction to authors is not a reason for calling an empty string an author error.

In fact the lack of a clear requirement in the spec that tells authors it’s OK to use the empty string for, e.g., `aria-activedescendant` is a reason for treating it as an error. As I pointed out in https://github.com/validator/validator/issues/557#issuecomment-335699935, at this point what actually is stated clearly in the spec at point of use for authors is that the value of `aria-activedescendant` must be a “Reference to the ID of another element in the same document” — with absolutely no mention there at point of use that it can instead by the empty string.

No author reading that part of the spec or any of the other parts of the spec that provide the requirements for states and properties is going to assume that it’s OK to use the empty string for any of them. Because that’s not what the spec is telling them.

Bottom line: If you believe the validator behavior for this should change, then I suggest you work to get changes made to the ARIA spec along the lines I’ve described.

**Comment by mcking65** [comment]
@sideshowbarker 

On Jan 11, 2019, @jnurthen merged a [commit](https://github.com/w3c/aria/commit/b5c237a9adf6e9d92a2be3a16df97d6067c59cf4) to fix w3c/aria#501 with the specific language you requested. It is  in [section 8.6 (State and Property Attribute Processing)](https://w3c.github.io/aria/#state_property_processing):

>Sometimes states and properties are present in the DOM but have a zero-length string ("") as their value. Authors MAY specify a zero-length string ("") for any supported (but not required) state or property.

Could we please have the validator updated accordingly?

It looks like I can't re-open this and remove the not bug label. Do you prefer a new issue?
